README.md
# mopidy-mpris

[![Latest PyPI version](https://img.shields.io/pypi/v/mopidy-mpris)](https://pypi.org/p/mopidy-mpris)
[![CI build status](https://img.shields.io/github/actions/workflow/status/mopidy/mopidy-mpris/ci.yml)](https://github.com/mopidy/mopidy-mpris/actions/workflows/ci.yml)
[![Test coverage](https://img.shields.io/codecov/c/gh/mopidy/mopidy-mpris)](https://codecov.io/gh/mopidy/mopidy-mpris)

[Mopidy](https://mopidy.com/) extension for controlling Mopidy through the MPRIS
D-Bus interface.

Mopidy-MPRIS supports the minimum requirements of the [MPRIS specification][1]
as well as the optional [Playlists interface][2]. The [TrackList interface][3]
is currently not supported.

[1]: https://specifications.freedesktop.org/mpris-spec/latest/
[2]: https://specifications.freedesktop.org/mpris-spec/latest/Playlists_Interface.html
[3]: https://specifications.freedesktop.org/mpris-spec/latest/Track_List_Interface.html


## Maintainer wanted

Mopidy-MPRIS is currently kept on life support by the Mopidy core developers.
It is in need of a more dedicated maintainer.

If you want to be the maintainer of Mopidy-MPRIS, please:

1. Make 2-3 good pull requests improving any part of the project.

2. Read and get familiar with all of the project's open issues.

3. Send a pull request removing this section and adding yourself as the
   "Current maintainer" in the "Credits" section below. In the pull request
   description, please refer to the previous pull requests and state that
   you've familiarized yourself with the open issues.

As a maintainer, you'll be given push access to the repo and the authority to
make releases to PyPI when you see fit.


## Installation

Install by running:

```sh
python3 -m pip install mopidy-mpris
```

See https://mopidy.com/ext/mpris/ for alternative installation methods.


## Configuration

No configuration is required for the MPRIS extension to work.

The following configuration values are available:

- `mpris/enabled`: If the MPRIS extension should be enabled or not.
  Defaults to ``true``.

- `mpris/bus_type`: The type of D-Bus bus Mopidy-MPRIS should connect to.
  Choices include `session` (the default) and `system`.
  
  
## Usage

Once Mopidy-MPRIS has been installed and your Mopidy server has been
restarted, the Mopidy-MPRIS extension announces its presence on D-Bus so that
any MPRIS compatible clients on your system can interact with it. Exactly how
you control Mopidy through MPRIS depends on which MPRIS client you use.


## Clients

The following clients have been tested with Mopidy-MPRIS.

### GNOME Shell builtin

- State: Not working
- Tested versions:
    - Ubuntu 18.10,
    - GNOME Shell 3.30.1-2ubuntu1,
    - Mopidy-MPRIS 2.0.0

GNOME Shell, which is the default desktop on Ubuntu 18.04 onwards, has a
builtin MPRIS client. This client seems to work well with Spotify's player,
but Mopidy-MPRIS does not show up here.

If you have any tips on what's missing to get this working, please open an
issue.

### gnome-shell-extensions-mediaplayer

- State: Working
- Tested versions:
    - Ubuntu 18.10,
    - GNOME Shell 3.30.1-2ubuntu1,
    - gnome-shell-extension-mediaplayer 63,
    - Mopidy-MPRIS 2.0.0
- Website: https://github.com/JasonLG1979/gnome-shell-extensions-mediaplayer

gnome-shell-extensions-mediaplayer is a quite feature rich MPRIS client
built as an extension to GNOME Shell. With the improvements to Mopidy-MPRIS
in v2.0, this extension works very well with Mopidy.

### gnome-shell-extensions-mpris-indicator-button

- State: Working
- Tested versions:
    - Ubuntu 18.10,
    - GNOME Shell 3.30.1-2ubuntu1,
    - gnome-shell-extensions-mpris-indicator-button 5,
    - Mopidy-MPRIS 2.0.0
- Website:
  https://github.com/JasonLG1979/gnome-shell-extensions-mpris-indicator-button/

gnome-shell-extensions-mpris-indicator-button is a minimalistic version of
gnome-shell-extensions-mediaplayer. It works with Mopidy-MPRIS, with the
exception of the play/pause button not changing state when Mopidy starts
playing.

If you have any tips on what's missing to get the play/pause button display
correctly, please open an issue.

### Ubuntu Sound Menu

- State: Unknown

Historically, Ubuntu Sound Menu was the primary target for Mopidy-MPRIS'
development. Since Ubuntu 18.04 replaced Unity with GNOME Shell, this is no
longer the case. It is currently unknown to what degree Mopidy-MPRIS works
with old Ubuntu setups.

If you run an Ubuntu setup with Unity and have tested Mopidy-MPRIS, please
open an issue to share your results.


## Advanced setups

### Running as a service

If you have input on how to best configure Mopidy-MPRIS when Mopidy is
running as a service, please add a comment to
[issue #15](https://github.com/mopidy/mopidy-mpris/issues/15).

### MPRIS on the system bus

You can set the `mpris/bus_type` config value to `system`. This will lead
to Mopidy-MPRIS making itself available on the system bus instead of the
logged in user's session bus.

> [!NOTE]
> Few MPRIS clients will try to access MPRIS devices on the system bus, so
> this will give you limited functionality. For example, media keys in
> GNOME Shell does not work with media players that expose their MPRIS
> interface on the system bus instead of the user's session bus.

The default setup will often not permit Mopidy to publish its service on the
D-Bus system bus, causing a warning similar to this in Mopidy's log:

```
MPRIS frontend setup failed (g-dbus-error-quark:
GDBus.Error:org.freedesktop.DBus.Error.AccessDenied: Connection ":1.3071"
is not allowed to own the service "org.mpris.MediaPlayer2.mopidy" due to
security policies in the configuration file (9))
```

To solve this, create the file
`/etc/dbus-1/system.d/org.mpris.MediaPlayer2.mopidy.conf` with the
following contents:

```xml
<!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN"
"http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">
<busconfig>
  <!-- Allow mopidy user to publish the Mopidy-MPRIS service -->
  <policy user="mopidy">
    <allow own="org.mpris.MediaPlayer2.mopidy"/>
  </policy>

  <!-- Allow anyone to invoke methods on the Mopidy-MPRIS service -->
  <policy context="default">
    <allow send_destination="org.mpris.MediaPlayer2.mopidy"/>
    <allow receive_sender="org.mpris.MediaPlayer2.mopidy"/>
  </policy>
</busconfig>
```

If you run Mopidy as another user than `mopidy`, you must
update `user="mopidy"` in the above file accordingly.

Once the file is in place, you must restart Mopidy for the change to take
effect.

To test the setup, you can run the following command as any user on the
system to play/pause the music:

```sh
dbus-send --system --print-reply \
  --dest=org.mpris.MediaPlayer2.mopidy \
  /org/mpris/MediaPlayer2 \
  org.mpris.MediaPlayer2.Player.PlayPause
```

### UPnP/DLNA with Rygel

[Rygel](https://wiki.gnome.org/Projects/Rygel) is an application that will
translate between Mopidy's MPRIS interface and UPnP. Rygel must be run on the
same machine as Mopidy, but will make Mopidy controllable by any device on the
local network that can control a UPnP/DLNA MediaRenderer.

The setup process is approximately as follows:

1.  Install Rygel.

    On Debian/Ubuntu:

    ```sh
    sudo apt install rygel
    ```

2.  Enable Rygel's MPRIS plugin.

    On Debian/Ubuntu, edit `/etc/rygel.conf`, find the `[MPRIS]`
    section, and change `enabled=false` to `enabled=true`.

3.  Start Rygel.

    To start it as the current user::

    ```sh
    systemctl --user start rygel
    ```

    To make Rygel start as the current user on boot::

    ```sh
    systemctl --user enable rygel
    ```

4.  Configure your system's firewall to allow the local network to reach
    Rygel. Exactly how is out of scope for this document.

5.  Start Mopidy with Mopidy-MPRIS enabled.

6.  If you view Rygel's log output with::

    ```sh
    journalctl --user -feu rygel
    ```

    You should see a log statement similar to::

    ```
    New plugin "org.mpris.MediaPlayer2.mopidy" available
    ```

6.  If everything went well, you should now be able to control Mopidy from a
    device on your local network that can control an UPnP/DLNA MediaRenderer,
    for example the Android app BubbleUPnP.

Alternatively, [upmpdcli combined with
Mopidy-MPD](https://docs.mopidy.com/en/latest/clients/upnp/) serves the same
purpose as this setup.



## Project resources

- [Source code](https://github.com/mopidy/mopidy-mpris)
- [Issues](https://github.com/mopidy/mopidy-mpris/issues)
- [Releases](https://github.com/mopidy/mopidy-mpris/releases)


## Development

### Set up development environment

Clone the repo using, e.g. using [gh](https://cli.github.com/):

```sh
gh repo clone mopidy/mopidy-mpris
```

Enter the directory, and install dependencies using [uv](https://docs.astral.sh/uv/):

```sh
cd mopidy-mpris/
uv sync
```

### Running tests

To run all tests and linters in isolated environments, use
[tox](https://tox.wiki/):

```sh
tox
```

To only run tests, use [pytest](https://pytest.org/):

```sh
pytest
```

To format the code, use [ruff](https://docs.astral.sh/ruff/):

```sh
ruff format .
```

To check for lints with ruff, run:

```sh
ruff check .
```

To check for type errors, use [pyright](https://microsoft.github.io/pyright/):

```sh
pyright .
```

### Adding features and fixing bugs

Mopidy-MPRIS has an extensive test suite, so the first step for all changes
or additions is to add a test exercising the new code. However, making the
tests pass doesn't ensure that what comes out on the D-Bus bus is correct. To
introspect this through the bus, there's a couple of useful tools.

### Browsing the MPRIS API with D-Feet

D-Feet is a graphical D-Bus browser. On Debian/Ubuntu systems it can be
installed by running:

```sh
sudo apt install d-feet
```

Then run the `d-feet` command. In the D-Feet window, select the tab
corresponding to the bus you run Mopidy-MPRIS on, usually the session bus.
Then search for "MediaPlayer2" to find all available MPRIS interfaces.

To get the current value of a property, double-click it. To execute a method,
double-click it, provide any required arguments, and click "Execute".

For more information on D-Feet, see the
[GNOME wiki](https://wiki.gnome.org/Apps/DFeet).

### Testing the MPRIS API with pydbus

To use the MPRIS API directly, start Mopidy, and then run the following in a
Python shell to use `pydbus` as an MPRIS client:

```pycon
>>> import pydbus
>>> bus = pydbus.SessionBus()
>>> player = bus.get('org.mpris.MediaPlayer2.mopidy', '/org/mpris/MediaPlayer2')
```

Now you can control Mopidy through the player object. To get properties from
Mopidy, run for example:

```pycon
>>> player.PlaybackStatus
'Playing'
>>> player.Metadata
{'mpris:artUrl': 'https://i.scdn.co/image/8eb49b41eeb45c1cf53e1ddfea7973d9ca257777',
  'mpris:length': 342000000,
  'mpris:trackid': '/com/mopidy/track/36',
  'xesam:album': '65/Milo',
  'xesam:albumArtist': ['Kiasmos'],
  'xesam:artist': ['Rival Consoles'],
  'xesam:discNumber': 1,
  'xesam:title': 'Arp',
  'xesam:trackNumber': 5,
  'xesam:url': 'spotify:track:7CoxEEsqo3XdvUsScRV4WD'}
>>>
```

To pause Mopidy's playback through D-Bus, run:

```pycon
>>> player.Pause()
>>>
```

For details on the API, please refer to the
[MPRIS specification](https://specifications.freedesktop.org/mpris-spec/latest/).

### Making a release

To make a release to PyPI, go to the project's [GitHub releases
page](https://github.com/mopidy/mopidy-mpris/releases)
and click the "Draft a new release" button.

In the "choose a tag" dropdown, select the tag you want to release or create a
new tag, e.g. `v0.1.0`. Add a title, e.g. `v0.1.0`, and a description of the changes.

Decide if the release is a pre-release (alpha, beta, or release candidate) or

CHANGELOG.rst
*********
Changelog
*********


v3.0.2 (2020-06-08)
===================

- Fix crash if cover art has no width.


v3.0.1 (2019-12-22)
===================

- Fix link in README.

- Fix tests by updating a mock.


v3.0.0 (2019-12-22)
===================

- Depend on final release of Mopidy 3.0.0.


v3.0.0rc1 (2019-11-12)
======================

- Require Mopidy >= 3.0.0a4, which required the following changes:

  - Stop using removed ``Album.images`` field.

  - Use ``uris`` instead of ``uri`` when calling ``core.tracklist.add()``.

- Require Python >= 3.7. No major changes required.

- Update project setup.


v2.0.0 (2018-12-07)
===================

Major feature release.

Dependencies
------------

- Require Mopidy >= 1.1.

- Replace python-dbus with python-pydbus.

Configuration
-------------

- Remove config value ``mpris/desktop_file``. It is marked as deprecated in
  the config schema, so it will be ignored if present in the config file.

Functionality
-------------

- Ordering of playlists by playlist modification time is no longer supported.

- Update UIs when playback options change: On the Mopidy event
  ``options_changed``, emit ``PropertiesChanged`` for ``LoopStatus``,
  ``Shuffle``, ``CanGoPrevious``, and ``CanGoNext``.

- Update UIs when playback is stopped: On the Mopidy event
  ``playback_state_changed``, emit ``PropertiesChanged`` for
  ``PlaybackStatus`` and ``Metadata``. (Fixes: #23)

- Update UIs when playlists are deleted: On the Mopidy event
  ``playlist_deleted``, emit``PropertiesChanged`` for ``PlaylistCount``.

- Update track name when stream title changes:

  - The ``Metadata`` property now uses ``core.playback.get_stream_title()`` as
    ``xesam:title`` if available.

  - On the Mopidy event ``stream_title_changed``, emit ``PropertiesChanged``
    for ``Metadata``.

- Control mixer mute through the volume control:

  - The ``Volume`` property is now ``0.0`` if the mixer is muted.

  - When setting the ``Volume`` property to a positive value, the mixer is
    unmuted.

  - On the Mopidy event ``mute_changed``, emit ``PropertiesChanged`` for
    ``Volume``.

- Fallback to get cover art from ``core.library.get_images()`` if
  ``track.album.images`` is blank.

- Do not expose Mopidy's desktop file through the ``DesktopEntry`` property. If
  we set this to "mopidy", the basename of "mopidy.desktop", some MPRIS clients
  will start a new Mopidy instance in a terminal window if one clicks outside
  the buttons of the UI. This is probably never what the user wants.

Internals
---------

- Improved documentation.

- Port tests to pytest.

- Replace all usage of Mopidy APIs deprecated as of Mopidy 2.2.


v1.4.0 (2018-04-10)
===================

- Remove dependency on python-indicate and libindicate, as it is deprecated and
  it no longer seems to no be necessary to send a startup notification with
  libindicate.

- Make tests pass with Mopidy >= 2.0.


v1.3.1 (2015-08-18)
===================

- Make tests pass with Mopidy >= 1.1.


v1.3.0 (2015-08-11)
===================

- No longer allow ``Quit()`` to shut down the Mopidy server process. Mopidy has
  no public API for Mopidy extensions to shut down the server.


v1.2.0 (2015-05-05)
===================

- Fix crash on seek event: Update ``seeked`` event handler to accept the
  ``time_position`` keyword argument. Recent versions of Mopidy passes all
  arguments to event handlers as keyword arguments, not positional arguments.
  (Fixes: #12)

- Fix crash on tracks longer than 35 minutes: The ``mpris:length`` attribute in
  the ``Metadata`` property is now typed to a 64-bit integer.

- Update ``Seek()`` implementation to only pass positive numbers to Mopidy, as
  Mopidy 1.1 is stricter about its input validation and no longer accepts seeks
  to negative positions.

- Add a hardcoded list of MIME types to the root interface
  ``SupportedMimeTypes`` property. This is a temporary solution to be able to
  play audio through UPnP using Rygel and Mopidy-MPRIS. Long term,
  mopidy/mopidy#812 is the proper solution. (Fixes: #7, PR: #11)

- Add a ``mpris/bus_type`` config value for making Mopidy-MPRIS connect to the
  D-Bus system bus instead of the session bus. (Fixes: #9, PR: #10)

- Update tests to pass with Mopidy 1.0.


v1.1.1 (2014-01-22)
===================

- Fix: Make ``OpenUri()`` work even if the tracklist is empty.


v1.1.0 (2014-01-20)
===================

- Updated extension API to match Mopidy 0.18.


v1.0.1 (2013-11-20)
===================

- Update to work with Mopidy 0.16 which changed some APIs.

- Remove redundant event loop setup already done by the ``mopidy`` process.


v1.0.0 (2013-10-08)
===================

- Moved extension out of the main Mopidy project.


mopidy_mpris/objects.py
from __future__ import unicode_literals

import base64
import logging
import os

import dbus
import dbus.service

from mopidy.core import PlaybackState


logger = logging.getLogger(__name__)

BUS_NAME = 'org.mpris.MediaPlayer2.mopidy'
OBJECT_PATH = '/org/mpris/MediaPlayer2'
ROOT_IFACE = 'org.mpris.MediaPlayer2'
PLAYER_IFACE = 'org.mpris.MediaPlayer2.Player'
PLAYLISTS_IFACE = 'org.mpris.MediaPlayer2.Playlists'


class MprisObject(dbus.service.Object):
    """Implements https://specifications.freedesktop.org/mpris-spec/2.2/"""

    properties = None

    def __init__(self, config, core):
        self.config = config
        self.core = core
        self.properties = {
            ROOT_IFACE: self._get_root_iface_properties(),
            PLAYER_IFACE: self._get_player_iface_properties(),
            PLAYLISTS_IFACE: self._get_playlists_iface_properties(),
        }
        bus_name = self._connect_to_dbus()
        dbus.service.Object.__init__(self, bus_name, OBJECT_PATH)

    def _get_root_iface_properties(self):
        return {
            'CanQuit': (True, None),
            'Fullscreen': (False, None),
            'CanSetFullscreen': (False, None),
            'CanRaise': (False, None),
            # NOTE Change if adding optional track list support
            'HasTrackList': (False, None),
            'Identity': ('Mopidy', None),
            'DesktopEntry': (self.get_DesktopEntry, None),
            'SupportedUriSchemes': (self.get_SupportedUriSchemes, None),
            # NOTE Return MIME types supported by local backend if support for
            # reporting supported MIME types is added
            'SupportedMimeTypes': (dbus.Array([
                dbus.String('audio/mpeg'),
                dbus.String('audio/x-ms-wma'),
                dbus.String('audio/x-ms-asf'),
                dbus.String('audio/x-flac'),
                dbus.String('audio/flac'),
                dbus.String('audio/l16;channels=2;rate=44100'),
                dbus.String('audio/l16;rate=44100;channels=2'),
                ], signature='s'), None),
        }

    def _get_player_iface_properties(self):
        return {
            'PlaybackStatus': (self.get_PlaybackStatus, None),
            'LoopStatus': (self.get_LoopStatus, self.set_LoopStatus),
            'Rate': (1.0, self.set_Rate),
            'Shuffle': (self.get_Shuffle, self.set_Shuffle),
            'Metadata': (self.get_Metadata, None),
            'Volume': (self.get_Volume, self.set_Volume),
            'Position': (self.get_Position, None),
            'MinimumRate': (1.0, None),
            'MaximumRate': (1.0, None),
            'CanGoNext': (self.get_CanGoNext, None),
            'CanGoPrevious': (self.get_CanGoPrevious, None),
            'CanPlay': (self.get_CanPlay, None),
            'CanPause': (self.get_CanPause, None),
            'CanSeek': (self.get_CanSeek, None),
            'CanControl': (self.get_CanControl, None),
        }

    def _get_playlists_iface_properties(self):
        return {
            'PlaylistCount': (self.get_PlaylistCount, None),
            'Orderings': (self.get_Orderings, None),
            'ActivePlaylist': (self.get_ActivePlaylist, None),
        }

    def _connect_to_dbus(self):
        bus_type = self.config['mpris']['bus_type']
        logger.debug('Connecting to D-Bus %s bus...', bus_type)

        if bus_type == 'system':
            bus = dbus.SystemBus()
        else:
            bus = dbus.SessionBus()

        bus_name = dbus.service.BusName(BUS_NAME, bus)
        logger.info('MPRIS server connected to D-Bus %s bus', bus_type)

        return bus_name

    def get_playlist_id(self, playlist_uri):
        # Only A-Za-z0-9_ is allowed, which is 63 chars, so we can't use
        # base64. Luckily, D-Bus does not limit the length of object paths.
        # Since base32 pads trailing bytes with "=" chars, we need to replace
        # them with an allowed character such as "_".
        encoded_uri = base64.b32encode(playlist_uri).replace('=', '_')
        return '/com/mopidy/playlist/%s' % encoded_uri

    def get_playlist_uri(self, playlist_id):
        encoded_uri = playlist_id.split('/')[-1].replace('_', '=')
        return base64.b32decode(encoded_uri)

    def get_track_id(self, tl_track):
        return '/com/mopidy/track/%d' % tl_track.tlid

    def get_track_tlid(self, track_id):
        assert track_id.startswith('/com/mopidy/track/')
        return track_id.split('/')[-1]

    # --- Properties interface

    @dbus.service.method(dbus_interface=dbus.PROPERTIES_IFACE,
                         in_signature='ss', out_signature='v')
    def Get(self, interface, prop):
        logger.debug(
            '%s.Get(%s, %s) called',
            dbus.PROPERTIES_IFACE, repr(interface), repr(prop))
        (getter, _) = self.properties[interface][prop]
        if callable(getter):
            return getter()
        else:
            return getter

    @dbus.service.method(dbus_interface=dbus.PROPERTIES_IFACE,
                         in_signature='s', out_signature='a{sv}')
    def GetAll(self, interface):
        logger.debug(
            '%s.GetAll(%s) called', dbus.PROPERTIES_IFACE, repr(interface))
        getters = {}
        for key, (getter, _) in self.properties[interface].iteritems():
            getters[key] = getter() if callable(getter) else getter
        return getters

    @dbus.service.method(dbus_interface=dbus.PROPERTIES_IFACE,
                         in_signature='ssv', out_signature='')
    def Set(self, interface, prop, value):
        logger.debug(
            '%s.Set(%s, %s, %s) called',
            dbus.PROPERTIES_IFACE, repr(interface), repr(prop), repr(value))
        _, setter = self.properties[interface][prop]
        if setter is not None:
            setter(value)
            self.PropertiesChanged(
                interface, {prop: self.Get(interface, prop)}, [])

    @dbus.service.signal(dbus_interface=dbus.PROPERTIES_IFACE,
                         signature='sa{sv}as')
    def PropertiesChanged(self, interface, changed_properties,
                          invalidated_properties):
        logger.debug(
            '%s.PropertiesChanged(%s, %s, %s) signaled',
            dbus.PROPERTIES_IFACE, interface, changed_properties,
            invalidated_properties)

    # --- Root interface methods

    @dbus.service.method(dbus_interface=ROOT_IFACE)
    def Raise(self):
        logger.debug('%s.Raise called', ROOT_IFACE)
        # Do nothing, as we do not have a GUI

    @dbus.service.method(dbus_interface=ROOT_IFACE)
    def Quit(self):
        logger.debug('%s.Quit called', ROOT_IFACE)
        # Do nothing, as we do not allow MPRIS clients to shut down Mopidy

    # --- Root interface properties

    def get_DesktopEntry(self):
        return os.path.splitext(os.path.basename(
            self.config['mpris']['desktop_file']))[0]

    def get_SupportedUriSchemes(self):
        return dbus.Array(self.core.uri_schemes.get(), signature='s')

    # --- Player interface methods

    @dbus.service.method(dbus_interface=PLAYER_IFACE)
    def Next(self):
        logger.debug('%s.Next called', PLAYER_IFACE)
        if not self.get_CanGoNext():
            logger.debug('%s.Next not allowed', PLAYER_IFACE)
            return
        self.core.playback.next().get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE)
    def Previous(self):
        logger.debug('%s.Previous called', PLAYER_IFACE)
        if not self.get_CanGoPrevious():
            logger.debug('%s.Previous not allowed', PLAYER_IFACE)
            return
        self.core.playback.previous().get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE)
    def Pause(self):
        logger.debug('%s.Pause called', PLAYER_IFACE)
        if not self.get_CanPause():
            logger.debug('%s.Pause not allowed', PLAYER_IFACE)
            return
        self.core.playback.pause().get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE)
    def PlayPause(self):
        logger.debug('%s.PlayPause called', PLAYER_IFACE)
        if not self.get_CanPause():
            logger.debug('%s.PlayPause not allowed', PLAYER_IFACE)
            return
        state = self.core.playback.state.get()
        if state == PlaybackState.PLAYING:
            self.core.playback.pause().get()
        elif state == PlaybackState.PAUSED:
            self.core.playback.resume().get()
        elif state == PlaybackState.STOPPED:
            self.core.playback.play().get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE)
    def Stop(self):
        logger.debug('%s.Stop called', PLAYER_IFACE)
        if not self.get_CanControl():
            logger.debug('%s.Stop not allowed', PLAYER_IFACE)
            return
        self.core.playback.stop().get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE)
    def Play(self):
        logger.debug('%s.Play called', PLAYER_IFACE)
        if not self.get_CanPlay():
            logger.debug('%s.Play not allowed', PLAYER_IFACE)
            return
        state = self.core.playback.state.get()
        if state == PlaybackState.PAUSED:
            self.core.playback.resume().get()
        else:
            self.core.playback.play().get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE, in_signature='x')
    def Seek(self, offset):
        logger.debug('%s.Seek called', PLAYER_IFACE)
        if not self.get_CanSeek():
            logger.debug('%s.Seek not allowed', PLAYER_IFACE)
            return
        offset_in_milliseconds = offset // 1000
        current_position = self.core.playback.time_position.get()
        new_position = current_position + offset_in_milliseconds
        if new_position < 0:
            new_position = 0
        self.core.playback.seek(new_position).get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE, in_signature='ox')
    def SetPosition(self, track_id, position):
        logger.debug('%s.SetPosition called', PLAYER_IFACE)
        if not self.get_CanSeek():
            logger.debug('%s.SetPosition not allowed', PLAYER_IFACE)
            return
        position = position // 1000
        current_tl_track = self.core.playback.current_tl_track.get()
        if current_tl_track is None:
            return
        if track_id != self.get_track_id(current_tl_track):
            return
        if position < 0:
            return
        if current_tl_track.track.length < position:
            return
        self.core.playback.seek(position).get()

    @dbus.service.method(dbus_interface=PLAYER_IFACE, in_signature='s')
    def OpenUri(self, uri):
        logger.debug('%s.OpenUri called', PLAYER_IFACE)
        if not self.get_CanControl():
            # NOTE The spec does not explictly require this check, but guarding
            # the other methods doesn't help much if OpenUri is open for use.
            logger.debug('%s.OpenUri not allowed', PLAYER_IFACE)
            return
        # NOTE Check if URI has MIME type known to the backend, if MIME support
        # is added to the backend.
        tl_tracks = self.core.tracklist.add(uri=uri).get()
        if tl_tracks:
            self.core.playback.play(tl_tracks[0]).get()
        else:
            logger.debug('Track with URI "%s" not found in library.', uri)

    # --- Player interface signals

    @dbus.service.signal(dbus_interface=PLAYER_IFACE, signature='x')
    def Seeked(self, position):
        logger.debug('%s.Seeked signaled', PLAYER_IFACE)
        # Do nothing, as just calling the method is enough to emit the signal.

    # --- Player interface properties

    def get_PlaybackStatus(self):
        state = self.core.playback.state.get()
        if state == PlaybackState.PLAYING:
            return 'Playing'
        elif state == PlaybackState.PAUSED:
            return 'Paused'
        elif state == PlaybackState.STOPPED:
            return 'Stopped'

    def get_LoopStatus(self):
        repeat = self.core.tracklist.repeat.get()
        single = self.core.tracklist.single.get()
        if not repeat:
            return 'None'
        else:
            if single:
                return 'Track'
            else:
                return 'Playlist'

    def set_LoopStatus(self, value):
        if not self.get_CanControl():
            logger.debug('Setting %s.LoopStatus not allowed', PLAYER_IFACE)
            return
        if value == 'None':
            self.core.tracklist.repeat = False
            self.core.tracklist.single = False
        elif value == 'Track':
            self.core.tracklist.repeat = True
            self.core.tracklist.single = True
        elif value == 'Playlist':
            self.core.tracklist.repeat = True
            self.core.tracklist.single = False

    def set_Rate(self, value):
        if not self.get_CanControl():
            # NOTE The spec does not explictly require this check, but it was
            # added to be consistent with all the other property setters.
            logger.debug('Setting %s.Rate not allowed', PLAYER_IFACE)
            return
        if value == 0:
            self.Pause()

    def get_Shuffle(self):
        return self.core.tracklist.random.get()

    def set_Shuffle(self, value):
        if not self.get_CanControl():
            logger.debug('Setting %s.Shuffle not allowed', PLAYER_IFACE)
            return
        if value:
            self.core.tracklist.random = True
        else:
            self.core.tracklist.random = False

    def get_Metadata(self):
        current_tl_track = self.core.playback.current_tl_track.get()
        if current_tl_track is None:
            return {'mpris:trackid': ''}
        else:
            (_, track) = current_tl_track
            metadata = {'mpris:trackid': self.get_track_id(current_tl_track)}
            if track.length:
                metadata['mpris:length'] = dbus.Int64(track.length * 1000)
            if track.uri:
                metadata['xesam:url'] = track.uri
            if track.name:
                metadata['xesam:title'] = track.name
            if track.artists:
                artists = list(track.artists)
                artists.sort(key=lambda a: a.name)
                metadata['xesam:artist'] = dbus.Array(
                    [a.name for a in artists if a.name], signature='s')
            if track.album and track.album.name:
                metadata['xesam:album'] = track.album.name
            if track.album and track.album.artists:
                artists = list(track.album.artists)
                artists.sort(key=lambda a: a.name)
                metadata['xesam:albumArtist'] = dbus.Array(
                    [a.name for a in artists if a.name], signature='s')
            if track.album and track.album.images:
                url = sorted(track.album.images)[0]
                if url:
                    metadata['mpris:artUrl'] = url
            if track.disc_no:
                metadata['xesam:discNumber'] = track.disc_no
            if track.track_no:
                metadata['xesam:trackNumber'] = track.track_no
            return dbus.Dictionary(metadata, signature='sv')

    def get_Volume(self):
        volume = self.core.playback.volume.get()
        if volume is None:
            return 0
        return volume / 100.0

    def set_Volume(self, value):
        if not self.get_CanControl():
            logger.debug('Setting %s.Volume not allowed', PLAYER_IFACE)
            return
        if value is None:
            return
        elif value < 0:
            self.core.playback.volume = 0
        elif value > 1:
            self.core.playback.volume = 100
        elif 0 <= value <= 1:
            self.core.playback.volume = int(value * 100)

    def get_Position(self):
        return self.core.playback.time_position.get() * 1000

    def get_CanGoNext(self):
        if not self.get_CanControl():
            return False
        current_tl_track = self.core.playback.current_tl_track.get()
        next_tl_track = self.core.tracklist.next_track(current_tl_track).get()
        return next_tl_track != current_tl_track

    def get_CanGoPrevious(self):
        if not self.get_CanControl():
            return False
        current_tl_track = self.core.playback.current_tl_track.get()
        previous_tl_track = (
            self.core.tracklist.previous_track(current_tl_track).get())
        return previous_tl_track != current_tl_track

    def get_CanPlay(self):
        if not self.get_CanControl():
            return False
        current_tl_track = self.core.playback.current_tl_track.get()
        next_tl_track = self.core.tracklist.next_track(current_tl_track).get()
        return current_tl_track is not None or next_tl_track is not None

    def get_CanPause(self):
        if not self.get_CanControl():
            return False
        # NOTE Should be changed to vary based on capabilities of the current
        # track if Mopidy starts supporting non-seekable media, like streams.
        return True

    def get_CanSeek(self):
        if not self.get_CanControl():
            return False
        # NOTE Should be changed to vary based on capabilities of the current
        # track if Mopidy starts supporting non-seekable media, like streams.
        return True

    def get_CanControl(self):
        # NOTE This could be a setting for the end user to change.
        return True

    # --- Playlists interface methods

    @dbus.service.method(dbus_interface=PLAYLISTS_IFACE)
    def ActivatePlaylist(self, playlist_id):
        logger.debug(
            '%s.ActivatePlaylist(%r) called', PLAYLISTS_IFACE, playlist_id)
        playlist_uri = self.get_playlist_uri(playlist_id)
        playlist = self.core.playlists.lookup(playlist_uri).get()
        if playlist and playlist.tracks:
            tl_tracks = self.core.tracklist.add(playlist.tracks).get()
            self.core.playback.play(tl_tracks[0]).get()

    @dbus.service.method(dbus_interface=PLAYLISTS_IFACE)
    def GetPlaylists(self, index, max_count, order, reverse):
        logger.debug(
            '%s.GetPlaylists(%r, %r, %r, %r) called',
            PLAYLISTS_IFACE, index, max_count, order, reverse)
        playlists = self.core.playlists.playlists.get()
        if order == 'Alphabetical':
            playlists.sort(key=lambda p: p.name, reverse=reverse)
        elif order == 'Modified':
            playlists.sort(key=lambda p: p.last_modified, reverse=reverse)
        elif order == 'User' and reverse:
            playlists.reverse()
        slice_end = index + max_count
        playlists = playlists[index:slice_end]
        results = [
            (self.get_playlist_id(p.uri), p.name, '')
            for p in playlists]
        return dbus.Array(results, signature='(oss)')

    # --- Playlists interface signals

    @dbus.service.signal(dbus_interface=PLAYLISTS_IFACE, signature='(oss)')
    def PlaylistChanged(self, playlist):
        logger.debug('%s.PlaylistChanged signaled', PLAYLISTS_IFACE)
        # Do nothing, as just calling the method is enough to emit the signal.

    # --- Playlists interface properties

    def get_PlaylistCount(self):
        return len(self.core.playlists.playlists.get())

    def get_Orderings(self):
        return [
            'Alphabetical',  # Order by playlist.name
            'Modified',      # Order by playlist.last_modified
            'User',          # Don't change order
        ]

    def get_ActivePlaylist(self):
        playlist_is_valid = False
        playlist = ('/', 'None', '')
        return (playlist_is_valid, playlist)
